2019.05.13 TC 리뷰사항 수정내역

Min-Su opened the pull request:
#6

Min-Su committed:
-- TC 리뷰사항 수정내역
예외처리에 대한 테스트 @Rule 어노테이션 사용 및 ExpectedException 사용으로 발생 될 Exception 및 메세지 검증
기능에 대한 시나리오에 대해 제대로 수행 되었는지 검증 코드 추가

monorisk commented on src/main/java/spring/ChangePasswordService.java:
```diff
@@ -14,7 +14,7 @@ public ChangePasswordService(MemberDao memberDao) {
     }
 
-    public void changePassword(String email, String oldPwd, String newPwd) {
+    public void changePassword(String email, String oldPwd, String newPwd) throws MemberNotFoundException {
```
음 `MemberNotFoundException.java` 는 문법적으로 try catch를 강제하지 않는 `unchecked exception` 입니다.

보통 이런 것들은 메서드 시그니처에 선언 하지 안는 편입니다.

이것도 어려운 주제이니 나중에 다시

https://stackoverflow.com/questions/824217/should-methods-that-throw-runtimeexception-indicate-it-in-method-signature

Min-Su replied:
https://plus4070.github.io/nhn%20entertainment%20devdays/Exception.html 참고하겠습니다.

monorisk commented on src/test/java/spring/MemberRegisterServiceTest.java:
```diff
@@ -1,58 +1,77 @@
     }
 
+    @Rule
+    public ExpectedException expectedException = ExpectedException.none();
```
자바에서 전통적인 멤버들 선언 순서

1. 상수
2. 변수
3. 생성자함수
4. 함수

monorisk replied:
```java
class Foo {
   public static final int BOO = 1;
   private static final int BAA = 2;

   public int gaa = 1;
   private int jaa = 2;

  public Foo() {
      
  }

  public Foo(int gaa) {
    this.gaa = gaa;
  } 

  public int doSomething() {
      // do something
      return 1;
  }

  private int doSomething2() {
     // also do something
    return 1;
  } 
}
```

monorisk commented on src/test/java/spring/MemberRegisterServiceTest.java:
```diff
@@ -1,58 +1,77 @@
-        req = new RegisterRequest();
-        memberRegSvc = new MemberRegisterService(memberDao);
+        expectedException.expect(DuplicateMemberException.class);
+        expectedException.expectMessage("dup email " + email);
```
문맥이 달라질 때는 줄바꿈

Min-Su replied:
줄바꿈은
expectedException.expectMessage("dup email " +
email); <- 이런걸 말하는 걸까요?

monorisk replied:
이런 댓글을 이제 봤음
```java
// expectedException이 무엇인지 정의하는 문맥
expectedException.expect(DuplicateMemberException.class);
expectedException.expectMessage("dup email " + email);

//  테스트에 사용할 member를 생성하는 문맥
Member member = new Member(email, password, name, LocalDateTime.now());

Min-Su committed:
-- ChangePasswordServiceTest, MemberRegisterServiceTest 시나리오 추가

Min-Su committed:
-- Member ChangePassword TC 추가

monorisk commented on src/test/java/spring/ChangePasswordServiceTest.java:
```diff
@@ -17,14 +17,47 @@
+        //when
+
+        memberDao.insert(member);
+        Member tmp_Member = memberDao.selectByEmail(email);
```
자바에서 _는 상수 빼고는 안씀

camel case 표기법 찾아봐바

```java
// 변수
Member tmpMember = memberDao.selectByEmail(email);

// 상수
static final String TEST_EMAIL = "testEmail@naver.com";
```

monorisk replied:
특정 언어군이나 집단군의 규약들을 어기면 코드 자체를 잘 신뢰 안하는 사람도 있으니 어지간하면 범용적인 수준까지는 지키는게 좋을듯

오라클에서 정한 엄격한 규칙은 자기네도 어길 때가 좀 있음
(물론 이게 camel case를 쓴다 아니다의 수준은 아니고 좀 더 복잡한 상황에서)

monorisk replied:
어라 `ChangePasswordService.java`에 대한 테스트가 있어야 하는데, `MemberDao.java`에 대한 테스트가 있음

위치 옮겨야 할듯

아래도 마찬가지고

monorisk replied:
사전작업 관점에서야 ChangePasswordService.java에서 사용하는 memberDao 빈을 초기화 시키고 테스트 조건에 맞게 데이터를 미리 넣어둘 수 있지만, 결과적으로 테스트 대상이 되는 것이 ChangePasswordService 빈이 되어야 함

monorisk commented on src/test/java/spring/ChangePasswordServiceTest.java:
```diff
@@ -17,14 +17,47 @@
+        //then
+
+        assertThat(tmp_Member).isNotNull();
+        assertThat(tmp_Member.getEmail()).isEqualTo(email);
```
사실 이런 류의 객체간의 동등성(equality)를 비교하는 것은 equals()를 오버라이드 하는 것이 최고겠지만 일단은 이건 나중에

monorisk replied:
지금은 이정도 비교로 괜찮다고 생각할 수 있지만, 정말 속성들이 같은지들은 비교인자들을 더 넣어야 할듯

이름, 비밀번호

등록일시는 나중에 비교하는 걸로

지금은 등록일시는 null인지 아닌지 정도만 보면 될듯?

monorisk commented on src/test/java/spring/MemberRegisterServiceTest.java:
```diff
@@ -41,17 +77,15 @@ public void setUp() {
-
     @Test
-    public void testRegist_이미_등록된_이메일이_있을_때() {
+    public void testRegist_이미_등록된_Email이_있을_때() {
```
요기에서 

`// when` 부분의 
```java
memberDao.insert(member);

req.setEmail(email);
req.setPassword(password);
req.setName(name);
```
는 

```java
memberRegSvc.regist(req);
```

테스트 조건을 위한 사전 작업이므로 `//given`의 영역임

monorisk commented on src/test/java/spring/MemberTest.java:
```diff
@@ -0,0 +1,57 @@
+
+        //then
+
+        assertThat(member.getPassword()).isEqualTo("5678");
```
비교 자체야 성공이긴 한데,

비교가 "5678"로 되어야 하는게 아니라 newPassword랑 되어야 할듯

이유라면, 테스트 코드도 유지 보수 비용이 들기 때문에 비용을 낮출 필요가 있음

```java
// 현재
assertThat(member.getPassword()).isEqualTo("5678");

// 좋은 케이스
assertThat(member.getPassword()).isEqualTo(newPassword);
```

이렇게 되면 newPassword의 값이 바뀌더라도, assertThat()구문의 값을 안바꿔도 되니까

Min-Su committed:
-- 2019-05-23 리뷰내용 수정사항
ChangPasswordServiceTest, MemberRegisterServiceTest, MemberTest 수정 완료
MemberDaoTest 추가
그 외 MemberInfoPrinterTest 추가 예정
